Mathlib/Geometry/Manifold/VectorField.lean

```diff
@@ -0,0 +1,317 @@
+
+variable {V W V₁ W₁ : Π (x : M'), TangentSpace I' x}
+variable {c : 𝕜} {m n : WithTop ℕ∞} {t : Set M'} {y₀ : M'}
+
+variable (I I') in
+/-- The pullback of a vector field under a map between manifolds, within a set `s`. -/
+def mpullbackWithin (f : M → M') (V : Π (x : M'), TangentSpace I' x) (s : Set M) (x : M) :
+    TangentSpace I x :=
+  (mfderivWithin I I' f s x).inverse (V (f x))
+
+variable (I I') in
+/-- The pullback of a vector field under a map between manifolds. -/
+def mpullback (f : M → M') (V : Π (x : M'), TangentSpace I' x) (x : M) :
+    TangentSpace I x :=
+  (mfderiv I I' f x).inverse (V (f x))
+
+lemma mpullbackWithin_apply :
+    mpullbackWithin I I' f V s x = (mfderivWithin I I' f s x).inverse (V (f x)) := rfl
+
+lemma mpullbackWithin_smul_apply :
+    mpullbackWithin I I' f (c • V) s x = c • mpullbackWithin I I' f V s x := by
+  simp [mpullbackWithin_apply]
+
+lemma mpullbackWithin_smul :
+    mpullbackWithin I I' f (c • V) s = c • mpullbackWithin I I' f V s := by
+  ext x
+  simp [mpullbackWithin_apply]
+
+lemma mpullbackWithin_add_apply :
+    mpullbackWithin I I' f (V + V₁) s x =
+      mpullbackWithin I I' f V s x + mpullbackWithin I I' f V₁ s x := by
+  simp [mpullbackWithin_apply]
+
+lemma mpullbackWithin_add :
+    mpullbackWithin I I' f (V + V₁) s =
+      mpullbackWithin I I' f V s + mpullbackWithin I I' f V₁ s := by
+  ext x
+  simp [mpullbackWithin_apply]
+
+lemma mpullbackWithin_neg_apply :
+    mpullbackWithin I I' f (-V) s x = - mpullbackWithin I I' f V s x := by
+  simp [mpullbackWithin_apply]
+
+lemma mpullbackWithin_neg :
+    mpullbackWithin I I' f (-V) s = - mpullbackWithin I I' f V s := by
+  ext x
+  simp [mpullbackWithin_apply]
+
+lemma mpullback_apply :
+    mpullback I I' f V x = (mfderiv I I' f x).inverse (V (f x)) := rfl
+
+lemma mpullback_add_apply :
+    mpullback I I' f (V + V₁) x = mpullback I I' f V x + mpullback I I' f V₁ x := by
+  simp [mpullback_apply]
+
+lemma mpullback_add :
+    mpullback I I' f (V + V₁) = mpullback I I' f V + mpullback I I' f V₁ := by
+  ext x
+  simp [mpullback_apply]
+
+lemma mpullback_neg_apply :
+    mpullback I I' f (-V) x = - mpullback I I' f V x := by
+  simp [mpullback_apply]
+
+lemma mpullback_neg :
+    mpullback I I' f (-V) = - mpullback I I' f V := by
+  ext x
+  simp [mpullback_apply]
+
+@[simp] lemma mpullbackWithin_univ : mpullbackWithin I I' f V univ = mpullback I I' f V := by
+  ext x
+  simp [mpullback_apply, mpullbackWithin_apply]
+
+lemma mpullbackWithin_eq_pullbackWithin {f : E → E'} {V : E' → E'} {s : Set E} :
+    mpullbackWithin 𝓘(𝕜, E) 𝓘(𝕜, E') f V s = pullbackWithin 𝕜 f V s := by
+  ext x
+  simp only [mpullbackWithin, mfderivWithin_eq_fderivWithin, pullbackWithin]
+  rfl
+
+@[simp] lemma mpullback_id {V : Π (x : M), TangentSpace I x} : mpullback I I id V = V := by
+  ext x
+  simp [mpullback]
+
+/-! ### Regularity of pullback of vector fields -/
+
+section
+
+variable [IsManifold I 2 M] [IsManifold I' 2 M'] [CompleteSpace E]
```

Comment:
Is completeness required? For which results? (Can you add a comment, if this is merely fore convenience?)